language/move-binary-format/src/control_flow_graph.rs

```diff
@@ -73,32 +93,104 @@ impl VMControlFlowGraph {
         }
 
         // Create basic blocks
-        let mut cfg = VMControlFlowGraph { blocks: Map::new() };
+        let mut blocks = Map::new();
         let mut entry = 0;
+        let mut exit_to_entry = Map::new();
         for pc in 0..code.len() {
-            let co_pc: CodeOffset = pc as CodeOffset;
+            let co_pc = pc as CodeOffset;
 
             // Create a basic block
-            if VMControlFlowGraph::is_end_of_block(co_pc, code, &block_ids) {
+            if Self::is_end_of_block(co_pc, code, &block_ids) {
+                let exit = co_pc;
+                exit_to_entry.insert(exit, entry);
                 let successors = Bytecode::get_successors(co_pc, code);
-                let bb = BasicBlock {
-                    entry,
-                    exit: co_pc,
-                    successors,
-                };
-                cfg.blocks.insert(entry, bb);
+                let bb = BasicBlock { exit, successors };
+                blocks.insert(entry, bb);
                 entry = co_pc + 1;
             }
         }
+        let blocks = blocks;
+        assert_eq!(entry, code_len);
+
+        // Determine traversal order
+        // build a DAG subgraph (remove the loop back edges)
+        let dag: Map<BlockId, Set<BlockId>> = blocks
+            .iter()
+            .map(|(id, block)| {
+                let id = *id;
+                let non_loop_continue_successors = block
+                    .successors
+                    .iter()
+                    // remove the loop back edges
+                    .filter(|successor| **successor > id)
+                    .copied()
+                    .collect();
+                (id, non_loop_continue_successors)
+            })
+            .collect();
```

Comment:
Some `debug_assert!`-guarded code that checks that this is actually a DAG would make me feel warmer/fuzzier...

Reply:
Thank you, I feel warmer/fuzzier.